### ISSUE TITLE
Possible special case optimisation for data that is initialised to 0?

### ISSUE THREAD

**Issue by mrdudz** [cc65/cc65#1202]
in my code i have 
```
static u16 stackptr = 0x0000;
```
which compiles to
```
.segment	"DATA"

_stackptr:
	.word	$0000
```
which is technically correct - but, is there a reason for why we couldnt we generally put this ("initialised data which is initialised to 0") into .bss instead of .data ? the benefit would be smaller binary.

i have no idea how hard this would be to do, obviously :)

**Comment by acqn** [comment]
> - but, is there a reason for why we couldnt **generally** put this ("initialised data which is initialised to 0") into .bss instead of .data ? the benefit would be smaller binary.

I suppose because:
- There is no guarantee that stuff in `.bss` will be initialized to 0. For instance, some of my projects are using a runtime library built without `zerobss.s` at all.
- Some actually compile with `-Oi` or a big `--CodeSize`, which suggests that time-efficiency is more important than space-efficiency to them.
- You can deliberately put something into `.bss` now. You need just remove its initializer.

> i have no idea how hard this would be to do, obviously :)

Maybe to implement a new compiler option and/or a new pragma.

**Comment by mrdudz** [comment]
> There is no guarantee that stuff in `.bss` will be initialized to 0

wait - isnt that pretty much defined as such?

**Comment by Marrin** [comment]
This should be optional and/or take care of only moving stuff from the standard DATA segment to BSS and not from other segments set as compiler option or via pragma.

**Comment by mrdudz** [comment]
of course, it should only happen with variables that are not explicitly put into a segment already

**Comment by oliverschmidt** [comment]
> There is no guarantee that stuff in .bss will be initialized to 0. For instance, some of my projects are using a runtime library built without zerobss.s at all.

I think there is.

> Some actually compile with -Oi or a big --CodeSize, which suggests that time-efficiency is more important than space-efficiency to them.

I've heard a rumour that a popular cc65 target machine has such a slow I/O system that decompressing stuff is faster than loading stuff ;-)) At least on that system it is supposed to be faster to initialize stuff with zero than loading zero stuff, isn't it?

**Comment by polluks** [comment]
> For instance, some of my projects are using a runtime library built without zerobss.s at all.
Well, if you modify crt0 you are always on your own.

**Comment by acqn** [comment]
In that case, why not formally put them into a `ZERODATA` segment and guarant that it will be zeroed at startup, instead of abusing the BSS segment?

https://cc65.github.io/doc/cc65.html#ss7.2:
> (the BSS segment is used to store **uninitialized** data)

**Comment by mrdudz** [comment]
still bss is guarenteed to be zeroed out :) if it wasnt, why would we have this zerobss function in the first place?

**Comment by jmr** [comment]
> `static u16 stackptr = 0x0000;`

`static u16 stackptr;` is also zero-initialized and that goes into bss. But, yes, seems reasonable to do this automatically.

**Comment by acqn** [comment]
Perhaps for putting "tentatively defined stuff" inside them. I consider that as an issue to be resolved.

**Comment by oliverschmidt** [comment]
> I should warn that we have had so many issues resulted from "re-ab-mis-using existing stuff".

Certainly true!

> In that case, why not formally put them into a ZERODATA segment and guarant that it will be zeroed at startup, instead of abusing the BSS segment?

Because then there would have to be another function beside `zerobss` to zero ZERODATA. Or `zerobss` would have to be parameterized and run twice. Both options would cause overhead likely not outweighed by the potential gains.

**Comment by oliverschmidt** [comment]
> static u16 stackptr; is also zero-initialized and that goes into bss. But, yes, seems reasonable to do this automatically.

Unfortunately I have no idea what you want to express. My understanding is that `static` only means that the variable in question isn't visible globally. It's still an uninitialized variable. Where else but BSS should it belong to? So what is done "automatically" here?

**Comment by oliverschmidt** [comment]
> Perhaps for putting "tentatively defined stuff" inside them. I consider that as an issue to be resolved.

Unfortunately I don't understand.

> EDIT: Just to clarify: the resolution with minimal coding efforts would be to fix the documentation about what BSS segments are (that the default is ZERODATA and the others are for uninitialized data).

Hm, now it's even less clear to me :-(

**Comment by acqn** [comment]
The default BSS segment is at the moment home for "stuff with static linkage that are not explicitly initialized", and there they will be all "zero-initialized" at startup. People want the compiler to automatically put "stuff with static linkage that are explicitly initialized to zero(es)" to be there too. I'd like to have the default one formally called `ZERODATA` (or `ZEROBSS`), or at least the documentation clarify that.

**Comment by jmr** [comment]
> > static u16 stackptr; is also zero-initialized and that goes into bss. But, yes, seems reasonable to do this automatically.
> 
> Unfortunately I have no idea what you want to express. My understanding is that `static` only means that the variable in question isn't visible globally. It's still an uninitialized variable. Where else but BSS should it belong to? 

Static variables without an explicit initializer are initialized to zero. https://port70.net/~nsz/c/c89/c89-draft.html#3.5.7

> So what is done "automatically" here?

"Automatically" would mean `static int x = 0;` goes in bss instead of making you change it to just `static int x;`.

**Comment by mrdudz** [comment]
I am surprised by the amount of discussion this generates :=) The core point of dispute seems to be that
a) strictly adhering the standard, an uninitalized objects value is indeterminate, and using it is undefined behaviour.
however there is
b) quasi standard behaviour in most toolchains is to initialize .bss with zero - and this is also the documented behaviour of cc65

IMHO it's a good thing b) is what it is, as that makes some things easier and shorter. and as a consequence, putting 0-initialized objects into .bss is a safe and trivial optimization. 

using an extra segment for those seems overkill to me (and it introduces extra overhead for very little gain), and i'd rather not do this at all.

**Comment by acqn** [comment]
Are there really "uninitialized objects" allowed by the C standard that are also candidates to be put into BSS in the first place?

**Comment by mrdudz** [comment]
*shrug*

**Comment by oliverschmidt** [comment]
Likely I'm completely missing the "interesting" aka esoteric aspect of this topic. What I know and understand is https://en.wikipedia.org/wiki/.bss#BSS_in_C

What the OP suggested is btw explicitly mentioned there:

_An implementation may also assign statically-allocated variables and constants initialized with a value consisting solely of zero-valued bits to the BSS section._

I certainly don't want to appear rude or something but I'm not interested in any way in PRs which change anything in the current BSS setup like introducing new segments, renaming segments or alike.

**Comment by mrdudz** [comment]
> I certainly don't want to appear rude or something but I'm not interested in any way in PRs which change anything in the current BSS setup like introducing new segments, renaming segments or alike.

it wasnt my intend to suggest that either - it just seemed an obvious and (likely) trivial optimization to me to put the zero initialised stuff into bss rather than data

**Comment by oliverschmidt** [comment]
...and at least the way I interpret the Wikipedia section I cited above seems to imply that this optimization is a well-known, reasonable one.

**Comment by acqn** [comment]
Obviously I don't know about C (or English) well enough to understand that the "_**un**intialized_" could mean "without explicit initializer" but NOT "_**not** initialized_". (Or I might have got this wrong!)

That ws all the reason I was arguing for renaming to `ZERODATA` here and about the cc65 documentation. Sorry guys I wasted your time.

> In C, statically-allocated objects without an explicit initializer are **initialized** to zero (for arithmetic types) or a null pointer (for pointer types).
>Hence, the BSS segment typically includes all **uninitialized** objects (both variables and constants) declared at file scope (i.e., outside any function) as well as uninitialized static local variables (local variables declared with the static keyword);

**Comment by oliverschmidt** [comment]
> [...] That was all the reason I was arguing for renaming to ZERODATA here and about the cc65 documentation.

Great to see that your root cause analysis skills do not only work for software bugs, but also for human misunderstandings ;-)

> Sorry guys I wasted your time.

You earned so many bonus lives in the recent weeks that I don't see you getting into this kind of trouble anytime soon!

**Comment by oliverschmidt** [comment]
So trying to summarize:

https://en.wikipedia.org/wiki/.bss#BSS_in_C says:

_An implementation may also assign statically-allocated variables and constants initialized with a value consisting solely of zero-valued bits to the BSS section._

And this is exactly what the OP proposes cc65 to do.

**Comment by jmr** [comment]
> a) strictly adhering the standard, an uninitalized objects value is indeterminate, and using it is undefined behaviour.

Just to make sure this is clear, uninitialized automatic variables have indeterminate value. Uninitialized static variables are 0.

> _An implementation may also assign statically-allocated variables and constants initialized with a value consisting solely of zero-valued bits to the BSS section._
>
> And this is exactly what the OP proposes cc65 to do.

I'm not sure the OP realized that just leaving off the initializer was an easy work-around. But, still a reasonable proposal.

**Comment by oliverschmidt** [comment]
> I'm not sure the OP realized that just leaving off the initializer was an easy work-around.

I am ;-)

**Comment by mrdudz** [comment]
> I'm not sure the OP realized that just leaving off the initializer was an easy work-around.

sure, i am. i just noticed this while hand optimizing a file in my project (where i actually wouldnt even benefit from this, because in that case even bss is statically linked in :))
